pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,224 @@
+	replicas := int32(1)
+
+	matchLbl := map[string]string{
+		"app.kubernetes.io/name":     "external-dns",
+		"app.kubernetes.io/instance": externalDNS.Name,
+	}
+
+	nodeSelectorLbl := map[string]string{
+		"kubernetes.io/os":               "linux",
+		"node-role.kubernetes.io/master": "",
+	}
+
+	depl := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      controller.ExternalDNSResourceName(externalDNS),
+			Namespace: namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: matchLbl,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: matchLbl,
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: serviceAccount.Name,
+					NodeSelector:       nodeSelectorLbl,
+				},
+			},
+		},
+	}
+
+	metricsPort := 7979
+	for i, zone := range externalDNS.Spec.Zones {
+		container, err := populateExternalDNSContainer(image, metricsPort+i, zone, externalDNS)
+		if err != nil {
+			return nil, err
+		}
+		depl.Spec.Template.Spec.Containers = append(depl.Spec.Template.Spec.Containers, *container)
+	}
+
+	return depl, nil
+}
+
+// WIP function for generating container specs for one container at a time.
+func populateExternalDNSContainer(image string, metricsPort int, zone string, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Container, error) {
+	name := "externaldns-" + zone
+
+	args := []string{
+		fmt.Sprintf("--metrics-address=127.0.0.1:%d", metricsPort),
+		fmt.Sprintf("--text-owner-id=externaldns-%s", externalDNS.Name),
+		fmt.Sprintf("--zone-id-filter=%s", zone),
+		"--policy=sync",
+		"--registry=txt",
+		"--log-level=debug",
+	}
+
+	provider, ok := providerStringTable[externalDNS.Spec.Provider.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported provider", externalDNS.Spec.Provider.Type)
+	}
+	args = append(args, fmt.Sprintf("--provider=%s", provider))
+
+	//TODO: Add provider credentials logic
+
+	source, ok := sourceStringTable[externalDNS.Spec.Source.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported source type", externalDNS.Spec.Source.Type)
+	}
+	args = append(args, fmt.Sprintf("--source=%s", source))
+
+	if externalDNS.Spec.Source.Namespace != nil && len(*externalDNS.Spec.Source.Namespace) > 0 {
+		args = append(args, fmt.Sprintf("--namespace=%s", *externalDNS.Spec.Source.Namespace))
+	}
+
+	if externalDNS.Spec.Source.Service != nil && len(externalDNS.Spec.Source.Service.ServiceType) > 0 {
+		serviceTypeFilter, publishInternal := "", false
+		for i, serviceType := range externalDNS.Spec.Source.Service.ServiceType {
+			serviceTypeFilter += string(serviceType)
+			// avoid having a trailing comma
+			if len(externalDNS.Spec.Source.Service.ServiceType)-1 > i {
+				serviceTypeFilter += ","
+			}
+			args = append(args, fmt.Sprintf("--service-type-filter=%s", serviceTypeFilter))
+
```

Comment:
Do you want one instance of the `service-type-filter` parameter, or several?

Reply:
It's possible to specify several but I think that it's a little more convenient to see them all on one line.

Reply:
Ah, now I see, yeah, it appends every time which doesn't make sense, good catch!